go/engine/merkle_audit.go

```diff
@@ -0,0 +1,251 @@
+	return []libkb.UIKind{}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *MerkleAudit) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{}
+}
+
+// Run starts the engine.
+// Returns immediately, kicks off a background goroutine.
+func (e *MerkleAudit) Run(m libkb.MetaContext) (err error) {
+	return RunEngine2(m, e.task)
+}
+
+func (e *MerkleAudit) Shutdown() {
+	e.task.Shutdown()
+}
+
+// randSeqno picks a random number between [low, high).
+func randSeqno(lo keybase1.Seqno, hi keybase1.Seqno) (keybase1.Seqno, error) {
+	rangeBig := big.NewInt(int64(hi - lo))
+	n, err := rand.Int(rand.Reader, rangeBig)
+	if err != nil {
+		return keybase1.Seqno(0), err
+	}
+	return keybase1.Seqno(n.Int64()) + lo, nil
+}
+
+var merkleAuditKey = libkb.DbKey{
+	Typ: libkb.DBMerkleAudit,
+	Key: "root",
+}
+
+func lookupMerkleAuditRetryFromState(m libkb.MetaContext) (*keybase1.Seqno, error) {
+	var state merkleAuditState
+	found, err := m.G().LocalDb.GetInto(&state, merkleAuditKey)
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		// Nothing found, no error
+		return nil, nil
+	}
+
+	// Can still be nil
+	return state.RetrySeqno, nil
+}
+
+func saveMerkleAuditState(m libkb.MetaContext, state merkleAuditState) error {
+	return m.G().LocalDb.PutObj(merkleAuditKey, nil, state)
+}
+
+func performMerkleAudit(m libkb.MetaContext, startSeqno keybase1.Seqno) error {
+	if m.G().ConnectivityMonitor.IsConnected(m.Ctx()) == libkb.ConnectivityMonitorNo {
+		m.CDebugf("MerkleAudit giving up offline")
+		return errAuditOffline
+	}
+
+	// Acquire the most recent merkle tree root
+	lastRoot := m.G().MerkleClient.LastRoot()
+	if lastRoot == nil {
+		m.CDebugf("MerkleAudit unable to retrieve the last root")
+		return errAuditNoLastRoot
+	}
+
+	// We can copy the pointer's value as it can only return nil if root == nil.
+	lastSeqno := *lastRoot.Seqno()
+
+	// Acquire the first root and calculate its hash
+	startRoot, err := m.G().MerkleClient.LookupRootAtSeqno(m, startSeqno)
+	if err != nil {
+		return err
+	}
+	startHash := startRoot.ShortHash()
+
+	// Traverse the merkle tree seqnos
+	currentSeqno := startSeqno + 1
+	step := 1
+	for {
+		// Proceed until the last known root
+		if currentSeqno > lastSeqno {
+			break
+		}
+
+		currentRoot, err := m.G().MerkleClient.LookupRootAtSeqno(m, currentSeqno)
+		if err != nil {
+			return err
+		}
+		currentHash := currentRoot.SkipToSeqno(startSeqno)
+		if currentHash == nil {
+			return libkb.NewClientMerkleSkipMissingError(
+				fmt.Sprintf("Root %d missing skip hash to %d", currentSeqno, startSeqno),
+			)
+		}
+
+		if !startHash.Eq(currentHash) {
+			// Warn the user about the possibility of the server tampering with the roots.
+			return libkb.NewClientMerkleSkipHashMismatchError(
+				fmt.Sprintf("Invalid skip hash from %d to %d", currentSeqno, startSeqno),
+			)
+		}
+
+		// We're doing this exponentially to make use of the skips.
+		currentSeqno += keybase1.Seqno(step)
+		step *= 2
```

Comment:
if this algo is correct to hit the relevant skips, then LGTM.

sorry for leading you to the rabbit hole with the tests.